python/common/org/python/types/Bytes.java

```diff
@@ -1251,10 +1251,79 @@ public Bytes rstrip() {
     }
 
     @org.python.Method(
-            __doc__ = "B.split(sep=None, maxsplit=-1) -> list of bytes\n\nReturn a list of the sections in B, using sep as the delimiter.\nIf sep is not specified or is None, B is split on ASCII whitespace\ncharacters (space, tab, return, newline, formfeed, vertical tab).\nIf maxsplit is given, at most maxsplit splits are done."
-    )
-    public org.python.Object split(java.util.List<org.python.Object> args, java.util.Map<java.lang.String, org.python.Object> kwargs, java.util.List<org.python.Object> default_args, java.util.Map<java.lang.String, org.python.Object> default_kwargs) {
-        throw new org.python.exceptions.NotImplementedError("bytes.split has not been implemented.");
+            __doc__ = "B.split(sep=None, maxsplit=None) -> list of bytes\n" +
+                    "\nReturn a list of the sections in B, using sep as the delimiter.\n" +
+                    "If sep is not specified or is None, B is split on ASCII whitespace\n" +
+                    "characters (space, tab, return, newline, formfeed, vertical tab).\n" +
+                    "If maxsplit is given, at most maxsplit splits are done.",
+            default_args = {"sep", "maxsplit"}
+    )
+    public org.python.Object split(org.python.Object sep, org.python.Object maxsplit) {
+        if (this.value.length == 0) {
+            if (sep == null) {
+                if (maxsplit == null || maxsplit instanceof org.python.types.Int) {
+                    return new org.python.types.List();
+                }
+                throw new org.python.exceptions.TypeError("'" + maxsplit.typeName() + "' object cannot be interpreted as an integer");
+            } else if (sep instanceof org.python.types.Bytes) {
+                if (maxsplit == null || maxsplit instanceof org.python.types.Int) {
+                    org.python.types.List result_list = new org.python.types.List();
+                    result_list.append(new org.python.types.Bytes("".getBytes()));
+                    return result_list;
+                }
+                throw new org.python.exceptions.TypeError("'" + maxsplit.typeName() + "' object cannot be interpreted as an integer");
+            } else {
+                if (org.Python.VERSION < 0x30500f0) {
+                    throw new org.python.exceptions.TypeError("'" + sep.typeName() + "' does not support the buffer interface");
+                } else {
+                    throw new org.python.exceptions.TypeError("a bytes-like object is required, not '" + sep.typeName() + "'");
+                }
+            }
+        }
+
+        byte[] bsep;
+        if (sep == null) {
+            bsep = " ".getBytes();
+        } else if (!(sep instanceof org.python.types.Bytes)) {
+            if (org.Python.VERSION < 0x30500f0) {
+                throw new org.python.exceptions.TypeError("'" + sep.typeName() + "' does not support the buffer interface");
```

Comment:
This is optional, but there are some bits of code that are repeating, like this -- what do you think about doing all these type checks in the beginning of the method (before the check on `this.value.length`) and throw any of these errors early?
This would make the code simpler and reduce the need for some of those test cases. :)